internal/provider/resource_eva_role.go

```diff
@@ -0,0 +1,270 @@
+	provider, diags := convertProviderType(in)
+
+	return roleProdiver{
+		provider: provider,
+	}, diags
+}
+
+type roleProdiver struct {
+	provider provider
+}
+
+type roleProviderTypeData struct {
+	ID                    types.Int64                 `tfsdk:"id"`
+	Name                  types.String                `tfsdk:"name"`
+	UserType              types.Int64                 `tfsdk:"user_type"`
+	Code                  types.String                `tfsdk:"code"`
+	ScopedFunctionalities []roleFunctionalityTypeData `tfsdk:"scoped_functionalities"`
+}
+
+type roleFunctionalityTypeData struct {
+	Functionality     types.String `tfsdk:"functionality"`
+	Scope             types.Int64  `tfsdk:"scope"`
+	RequiresElevation types.Bool   `tfsdk:"requires_elevation"`
+}
+
+func (s roleProviderTypeData) getListOfFunctionalities() []eva.RoleFunctionality {
+	var scopedFunctionalities []eva.RoleFunctionality
+
+	for _, scopedFunctionality := range s.ScopedFunctionalities {
+		scopedFunctionalities = append(scopedFunctionalities, eva.RoleFunctionality{
+			Functionality:     scopedFunctionality.Functionality.Value,
+			Scope:             scopedFunctionality.Scope.Value,
+			RequiresElevation: scopedFunctionality.RequiresElevation.Value,
+		})
+	}
+
+	return scopedFunctionalities
+}
+
+func (s roleProviderTypeData) setListOfFunctionalities(scopedFunctionalities []eva.RoleFunctionality) {
+	for _, scopedFunctionality := range scopedFunctionalities {
+		s.ScopedFunctionalities = append(s.ScopedFunctionalities, roleFunctionalityTypeData{
+			Functionality:     types.String{Value: scopedFunctionality.Functionality},
+			Scope:             types.Int64{Value: scopedFunctionality.Scope},
+			RequiresElevation: types.Bool{Value: scopedFunctionality.RequiresElevation},
+		})
+	}
+}
+
+func (r roleProdiver) Create(ctx context.Context, req tfsdk.CreateResourceRequest, resp *tfsdk.CreateResourceResponse) {
+	var data roleProviderTypeData
+
+	diags := req.Config.Get(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	createdRole, createRoleErr := r.provider.evaClient.CreateRole(ctx, eva.CreateRoleRequest{
+		Name:     data.Name.Value,
+		UserType: data.UserType.Value,
+		Code:     data.Code.Value,
+	})
+
+	if createRoleErr != nil {
+		resp.Diagnostics.AddError("Creating role unit failed.", fmt.Sprintf("Unable to create role, got error: %s", createRoleErr))
+		return
+	}
+
+	data.ID = types.Int64{Value: createdRole.ID}
+
+	_, attachPermissionsToRoleErr := r.provider.evaClient.AttachFunctionalitiesToRole(ctx, eva.AttachFunctionalitiesToRoleRequest{
+		RoleID:                data.ID.Value,
+		ScopedFunctionalities: data.getListOfFunctionalities(),
+	})
+
+	if attachPermissionsToRoleErr != nil {
+		resp.Diagnostics.AddError("Creating role permissions failed.", fmt.Sprintf("Unable to create role permisions, got error: %s", attachPermissionsToRoleErr))
+		return
```

Comment:
Here you are returning, however the `role` was successfully created. That would lead to orphaned roles in case attachment fails. Not sure this is desired behaviour.

Reply:
Okay, maybe we need to revert the changes (this means deleting the newly created rule)

Reply:
There are things I think we could do:
1. Log an error saying "roles is not fully configured, please run the plan again" but **not** doing the `return`. That way the state is still successfully stored and the ID is not lost.
2. Deleting like you mentioned, however this also has a chance of failing as well 🙈